operator/rediscontroller/controller_test.go

```diff
@@ -0,0 +1,339 @@
+	ts.Assert().NoError(err)
+
+	instance = &exoscalev1.Redis{}
+	ts.FetchResource(types.NamespacedName{Name: mg.Name, Namespace: mg.Namespace}, instance)
+
+	ts.Assert().Equal(instance.Spec.ForProvider.Maintenance.DayOfWeek, oapi.DbaasServiceMaintenanceDowFriday)
+
+	ready := false
+	synced := false
+	for _, condition := range instance.Status.Conditions {
+		if condition.Type == xpv1.TypeReady {
+			ts.Assert().Equal(corev1.ConditionTrue, condition.Status)
+			ts.Assert().Equal(xpv1.ReasonAvailable, condition.Reason)
+			ready = true
+		} else if condition.Type == xpv1.TypeSynced {
+			ts.Assert().Equal(corev1.ConditionTrue, condition.Status)
+			ts.Assert().Equal(xpv1.ReasonReconcileSuccess, condition.Reason)
+			synced = true
+		}
+	}
+	ts.Assert().True(ready && synced, "not ready & synced (ready: %t / synced: %t)", ready, synced)
+
+	ts.ExoClientMock.AssertExpectations(ts.T())
+}
+
+func (ts *RedisControllerTestSuite) TestDelete() {
+	ts.mockCaCertificateResponse()
+
+	name := "test-delete"
+	settings := map[string]any{
+		"maxmemory_policy": "noeviction",
+	}
+
+	mg := newRedisInstance(name, settings)
+	mg.Status = exoscalev1.RedisStatus{
+		ResourceStatus: xpv1.ResourceStatus{},
+		AtProvider:     exoscalev1.RedisObservation{},
+	}
+	redisResponse := ts.getRedisResponse(mg, settings)
+
+	ts.EnsureResources(mg)
+
+	// reply once with a still existing redis, to allow deletion to happen.
+	ts.ExoClientMock.On("GetDbaasServiceRedisWithResponse", mock.Anything, oapi.DbaasServiceName(name)).
+		Return(redisResponse, nil).Twice()
+	ts.ExoClientMock.On("GetDbaasServiceRedisWithResponse", mock.Anything, oapi.DbaasServiceName(name)).
+		Return(nil, exoscaleapi.ErrNotFound)
+
+	// external resource is up-to-date
+	_, err := ts.reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: mg.Name, Namespace: mg.Namespace}})
+	ts.Assert().NoError(err)
+
+	ts.ExoClientMock.On("DeleteDbaasServiceWithResponse", mock.Anything, name).
+		Return(&oapi.DeleteDbaasServiceResponse{Body: []byte{}}, nil).
+		Once()
+
+	ts.DeleteResources(mg)
+
+	// requested deletion of external resource
+	_, err = ts.reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: mg.Name, Namespace: mg.Namespace}})
+	ts.Assert().NoError(err)
+
+	// removes finalizer
+	_, err = ts.reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: mg.Name, Namespace: mg.Namespace}})
+	ts.Assert().NoError(err)
+
+	instance := &exoscalev1.Redis{}
+	err = ts.Client.Get(ts.Context, types.NamespacedName{Name: mg.Name, Namespace: mg.Namespace}, instance)
+	ts.Assert().True(apierrors.IsNotFound(err))
+
+	ts.ExoClientMock.AssertExpectations(ts.T())
+}
+
+// reconcile waits reconcileWaitPeriod before running a reconciliation.
+func (ts *RedisControllerTestSuite) reconcile(req reconcile.Request) (reconcile.Result, error) {
+	ts.T().Helper()
+	time.Sleep(ts.reconcileWaitPeriod)
+	return ts.reconciler.Reconcile(ts.Context, req)
+}
```

Comment:
Why is the time delay needed?

Reply:
after reconcile there's often the case that the resource status gets updated. This gives it time to persist the changes. 
If removed:

```
        	            	cannot update managed resource status: Operation cannot be fulfilled on redis.exoscale.crossplane.io "test-create": the object has been modified; please apply your changes to the latest version and try again
```

It's a bit ugly and I can imagine that it may be different depending on the machine the test is executed on. Not sure how to avoid it without spending too much time on it. 

Do you have an idea?

Reply:
> after reconcile there's often the case that the resource status gets updated.

any idea who/what does this? the manager in the background?

> please apply your changes to the latest version and try again

that seems to be a pretty common error in Crossplane. We get these constantly in prod as well and we were unable to get rid of them yet... It's entirely possible that this is a bug in Crossplane runtime itself and not in our code.

Reply:
I guess that's just the normal [`Reconcile`](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go#L637) which does it. Usually crossplane seems to do two or more reconciles for one change to make sure everything is done and safe. Not sure if they do too many updates or what the issue is there. But basicly every return in that function at least sets conditions, which may or may not always lead to a status update.

Reply:
yeah... conditions are set but the status not updated. Which indicates that a higher abstraction (manager?) does the status updates.
I don't like this, since the tests contain now delays and that makes the test unstable as it can easily fail in other environments and introduce hard-to-find bugs in test code...
Anyway, let's leave it as-is with a comment
```suggestion
// reconcile waits reconcileWaitPeriod before running a reconciliation.
func (ts *RedisControllerTestSuite) reconcile(req reconcile.Request) (reconcile.Result, error) {
	ts.T().Helper()
	// wait a bit so that status updates are done in the background (by the manager?)
	time.Sleep(ts.reconcileWaitPeriod)
	return ts.reconciler.Reconcile(ts.Context, req)
}
```

Reply:
mh the status is [always updated too](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go#L675):
```
		return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
```

^ that's btw the reason why I don't like the `errors.Wrap()`. It makes code harder to read ;)

Agreeing on the not liking this. I'd prefer to find the root cause and fix it. We can discuss if I should invest a day or two into that I guess? Let's talk later. 
I added the comment :+1: